src/constructors/lattices.c

```diff
@@ -0,0 +1,299 @@
+    }
+
+    IGRAPH_CHECK(triangle_lattice(graph, directed, mutual, &row_lengths_vector, &row_start_vector));
+
+    igraph_vector_int_destroy(&row_lengths_vector);
+    igraph_vector_int_destroy(&row_start_vector);
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t triangle_lattice_hex_shape(
+    igraph_t *graph, igraph_integer_t size_x, igraph_integer_t size_y,
+    igraph_integer_t size_z, igraph_bool_t directed, igraph_bool_t mutual
+) {
+    igraph_integer_t row_count = size_y + size_z - 1;
+    igraph_vector_int_t row_lengths_vector;
+    igraph_vector_int_t row_start_vector;
+    igraph_integer_t i;
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&row_lengths_vector, row_count);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&row_start_vector, row_count);
+
+    igraph_integer_t row_length = size_x;
+    igraph_integer_t row_start = size_y - 1;
+    igraph_integer_t first_threshold = MIN(size_y - 1, size_z - 1);
+    igraph_integer_t second_threshold = MAX(size_y - 1, size_z - 1);
+    igraph_integer_t sgn_flag = size_y < size_z ? 0 : -1;
+
+    for (i = 0; i < row_count; i++) {
+        VECTOR(row_lengths_vector)[i] = row_length;
+        VECTOR(row_start_vector)[i] = row_start;
+
+        if (i < first_threshold) {
+            row_length++;
+            row_start--;
+        } else if (i < second_threshold) {
+            row_start += sgn_flag;
+        } else {
+            row_length--;
+        }
+    }
+
+    IGRAPH_CHECK(triangle_lattice(graph, directed, mutual, &row_lengths_vector, &row_start_vector));
+
+    igraph_vector_int_destroy(&row_lengths_vector);
+    igraph_vector_int_destroy(&row_start_vector);
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \ingroup generators
+ * \function igraph_triangle_lattice
+ * \brief A triangle lattice with the given shape.
+ *
+ * Creates a triangle lattice whose vertices have the form (i, j) for non-negative integers i and j
+ * and (i, j) is generally connected with (i + 1, j), (i, j + 1), and (i - 1, j + 1).
+ *
+ * </para><para>
+ * The vertices of the resulting graph are ordered lexicographically with the 2nd coordinate being
+ * more significant, e.g., (i, j) < (i + 1, j) and (i + 1, j) < (i, j + 1)
+ *
+ * \param graph An uninitialized graph object.
+ * \param dims Integer vector, defines the shape of the lattice. (Below the "edge length"s are in terms of graph theoretical path lengths.)
+ *        If \p dims is of length 1, the resulting lattice has a triangular shape
+ *        where each side of the triangle contains \c "dims[0]" vertices.
+ *        If \p dims is of length 2, the resulting lattice has a
+ *        "quasi rectangular" shape with the sides containing \c "dims[0]" and
+ *        \c "dims[1]" vertices, respectively.
+ *        If \p dims is of length 3, the resulting lattice has a hexagonal shape
+ *        where the sides of the hexagon contain \c "dims[0]", \c "dims[1]" and
+ *        \c "dims[2]" vertices.
```

Comment:
Let's change `\c "dims[0]"` to `<code>dims[0]</code>`. Necessary because of the limitations of `\c`.